### PR TITLE
eval_everywhere & safe_eval_everywhere Preview

### PR DESCRIPTION
**Pull request by linearregression**
*eval_everywhere 
Just like rpc's  verison apply(Nodes, Module, Function) on the specified nodes. Pure fire and pray.
Always return abcast. Nodes is [] or [::node()]
eval_everywhere(Nodes, M, F)   <-- rpc left this out 
eval_everywhere(Nodes, M, F, A),
eval_everywhere(Nodes, M, F, A, SendTO)

*safe_eval_everywhere (rpc does not have  equivalent version)
apply(Nodes, Module, Function) on the specified nodes. Returns a list of nodes with status true or {badrpc, Reason} or {badtcp, Reason}.
e.g. If SLAVE and SLAVE1 are present to receive messages. Error/Sucess Semantics same as safe_cast.
e.g.
[{?SLAVE,true}, {?SLAVE1,true},{?FAKE_NODE,{badrpc,nodedown}}]
Apply on empty nodes return []. 
safe_eval_everywhere(Nodes, M, F),
safe_eval_everywhere(Nodes,M,F,A),
safe_eval_everywhere(Nodes, M, F, A, SendTO)

Note:
* on throw,it will trigger a {nocatch,throwXup}, ....} on the target node. 

Tests:
* Since the fire and pray nature, we don't know if target system received anything. The original idea was to store something uniquely recognizable for that run and use rpc:call to get the result back to check.However, rpc:call always get there earlier than gen_rpc. So now it is no better than just like checking the return signature like gen_rpc:cast has been doing.
* The test server has dbg trace on for some eye balling (which is subopt).

* going to replace test with something simpler

### CODE REVIEW

**Comment by priestjim** [comment]
Can you please rebase off of develop first?

**Comment by linearregression** [comment]
Looks like they updated the "Files changed"  section.

**Comment by linearregression** [comment]
Hi, got the chance to check it out yet?

File: src/gen_rpc_client.erl [hunk 1]
```diff
@@ -148,6 +165,21 @@ safe_cast(Node, M, F, A, SendTO) when is_atom(Node), is_atom(M), is_atom(F), is_
...
+
+%% Safe evaluate Module:Function:Arguments on implicit connected nodes. Custom sender timeout.
+safe_eval_everywhere(Nodes, M, F, A) ->
+    safe_eval_everywhere(Nodes, M, F, A, 'undefined').
+
+
+%% Safe evaluate Module:Function:Arguments on custom list of nodes.
+safe_eval_everywhere(Nodes, M, F, A, SendTO) ->
+    ok = lager:debug("function=safe_eval_everywhere_mfa_to event=eval_on_nodes nodes=\"~p\"", [Nodes]),
+    [{Node, safe_cast(Node, M, F, A, SendTO)} || Node <- Nodes].
```
Comment (priestjim) [inline src/gen_rpc_client.erl]: Why return a tuple here? The function is supposed to return just the results in proper order. Also, per the usual "multicall" paradigm, this should return {true, BadNodes} where BadNodes are the nodes with badtcp/badrpc results

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    [{?SLAVE,true}, {?SLAVE1,true}] = gen_rpc:safe_eval_everywhere(ConnectedNodes, erlang, throw, ['throwXup']),
+    ok = ct:pal("[erlang:throw only]. Verify the crash log from ct. You should see {{nocatch,throwXup}, ....} on the target node").
+
+safe_eval_everywhere_mfa_timeout_on_nodes(_Config) ->
+    ok = ct:pal("Testing [eval_everywhere_mfa_timeout_on_nodes]"),
+    ConnectedNodes = [?SLAVE, ?SLAVE1],
+    [{?SLAVE,true}, {?SLAVE1,true}] = gen_rpc:safe_eval_everywhere(ConnectedNodes, erlang, throw, ['throwXup']),
+    ok = ct:pal("erlang:throw only]. Verify the crash log from ct. You should see {{nocatch,throwXup}, ....} on the target node").
+
+client_inactivity_timeout(_Config) ->
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Why are these here? Aren't they part of the remote node suite?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    true = erlang:is_process_alive(whereis(gen_rpc_client_sup)).
+
+eval_everywhere_mfa_one_node(_Config) ->
+    ok = ct:pal("Testing [eval_everywhere_mfa_one_node]"),
+    ConnectedNodes = [?SLAVE],
+    Data = make_data('eval_everywhere_mfa_one_node'),
+    abcast = gen_rpc:eval_everywhere(ConnectedNodes, test_app_server, ping),
+    abcast = gen_rpc:eval_everywhere(ConnectedNodes, test_app_server, store, [Data]).
+
+eval_everywhere_mfa_on_nodes(_Config) ->
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: _on_nodes -> _multiple_nodes

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    ok = ct:pal("Testing [eval_everywhere_mfa_on_nodes]"),
+    ConnectedNodes = [?SLAVE, ?SLAVE1],
+    Data = make_data('eval_everywhere_mfa_on_nodes'),
+    abcast = gen_rpc:eval_everywhere(ConnectedNodes, test_app_server, store, [Data]).
+
+eval_everywhere_mfa_on_nodes_TO(_Config) ->
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: What does this test test?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    %% Lookup the client named process, shouldn't be there
+    [] = supervisor:which_children(gen_rpc_acceptor_sup),
+    %% The server supervisor should have no children
+    [] = supervisor:which_children(gen_rpc_server_sup).
+
+%%% ===================================================
+%%% Auxiliary functions for test cases
+%%% ===================================================
+start_slaves() ->
+    ok = start_slave(?SLAVE_NAME, ?SLAVE),
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Better name the slaves in this situation SLAVE1, SLAVE2

**Comment by priestjim** [comment]
What you can do is use call to fetch the state from the test gen_server and see if the call has been evaluated. But you need to flush it for every test.

**Comment by linearregression** [comment]
* ditched test server. Too many auxiliary test modules.
* master node will trigger slave nodes to ping him with test case specific message.
* master node relay the event to test script to verify

**Comment by linearregression** [comment]
* ditch test sever. Too many auxiliary test modules.
* master node trigger slaves nodes to ping him back with test case unique message
* master node relay the results back to test script to verify results

File: src/gen_rpc_client.erl [hunk 6]
```diff
@@ -389,3 +422,17 @@ merge_timeout_values(SRecvTO, undefined, _SSendTO, USendTO) ->
     {SRecvTO, USendTO};
 merge_timeout_values(_SRecvTO, URecvTO, _SSendTO, USendTO) ->
     {URecvTO, USendTO}.
+
+%% Transform result for safe_eval_everywhere to look like multicall
+transform_result(Result, Nodes) ->
```
Comment (priestjim) [inline src/gen_rpc_client.erl]: I think this part can be improved. We only care about badrpc/badtcp so why not a list comprehension with filtering the bad nodes like `{_, _} = X` ?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+-author("Panagiotis Papadomitsos <pj@ezgr.net>").
+
+%%% CT Macros
+-include_lib("test/gen_rpc/include/ct.hrl").
+
+%%% Common Test callbacks
+-export([all/0, init_per_suite/1, end_per_suite/1, init_per_testcase/2, end_per_testcase/2]).
+
+%%% Testing functions
+-export([supervisor_black_box/1
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: In the interest of keeping the same coding style, can you move the commas on the upper line like in all other modules?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+end_per_testcase(_OtherTest, Config) ->
+    ok = stop_slaves(),
+    Config.
+
+
+%%% ===================================================
+%%% Test cases
+%%% ===================================================
+%% Test supervisor's status
+supervisor_black_box(_Config) ->
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Why here again?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+
+clean_process(Name, Reason) ->
+    Pid = whereis(Name),
+    true = kill_it(Pid, Name, Reason),
+    ok.
+
+kill_it(undefined, _Name, _Reason) -> true;
+kill_it(Pid, Name, Reason) ->
+    true = exit(Pid, Reason),
+    true = unregister(Name).
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: When a process exits, it automatically unregisters.

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    Msg = Name = 'evalmfamul', 
+    TestPid = self(),
+    ok = clean_process(Name, 'normal'),
+    Pid = spawn_listener2(?SLAVE1, ?SLAVE2, Name, TestPid, 2),
+    true = register(Name, Pid),
+    abcast = gen_rpc:eval_everywhere(ConnectedNodes, 'gen_rpc_test_helper', ping, [{?NODE, Name, Msg}]),
+    {ok, passed} = wait_for_reply(?SLAVE1, ?SLAVE2),
+    ok.
+
+eval_everywhere_mfa_multiple_nodes_TO(_Config) ->
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: TO -> timeout :)

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+    [true, []] = gen_rpc:safe_eval_everywhere(ConnectedNodes, erlang, throw, ['throwXup']),
+    ok = ct:pal("[erlang:throw only]. Verify the crash log from ct. You should see {{nocatch,throwXup}, ....} on the target node").
+
+safe_eval_everywhere_mfa_timeout_multiple_nodes(_Config) ->
+    ok = ct:pal("Testing [eval_everywhere_mfa_timeout_multiple_nodes]"),
+    ConnectedNodes = [?SLAVE1, ?SLAVE2],
+    [true, []] = gen_rpc:safe_eval_everywhere(ConnectedNodes, erlang, throw, ['throwXup']),
+    ok = ct:pal("erlang:throw only]. Verify the crash log from ct. You should see {{nocatch,throwXup}, ....} on the target node").
+
+client_inactivity_timeout(_Config) ->
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Why here again?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+stop_slaves() ->
+    Slaves = [?SLAVE1, ?SLAVE2],
+    [begin 
+        ok = slave:stop(Node)
+     end || Node <- Slaves],
+    ok = ct:pal("Slaves stopped").
+
+%% This is the middleman process listening for messages from slave nodes
+%% Then relay back to test case Pid for check.
+spawn_listener(_, _, _, Count) when Count =< 0-> ok;
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Shouldn't this and wait_for_reply go to the helper module?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+                Else -> ok = ct:pal("Unknown Message: \"~p\"", [Else]),
+                        TestPid ! Else,
+                        loop(Node1, Node2, Name, TestPid, Count)
+          after
+                5000 ->
+                        ok = ct:pal("pong timeout", []),
+                        {error, pong_timeout}
+          end.
+
+wait_for_reply(Node1, Node2) ->
```
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: I may change the formatting here after merge, I hope it's ok.

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 11 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: are u talking about the server and client inactivity tests?
If so they are dups and can be removed.

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 13 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Hi, Do whatever you want. You are the ultimate gatekeeper of this project. :)

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 11 above)
Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Yes! Why would have them duped anyway?

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 7 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: done

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 8 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: removed.

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 9 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: fixed

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 10 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: done

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 11 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: removed

File: test/gen_rpc/multi_rpc_functional_SUITE.erl (hunk 12 above)
Comment (linearregression) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: Only this test uses this. Do you like to put this int helper?
It may be useful for other tests later on?

Comment (priestjim) [inline test/gen_rpc/multi_rpc_functional_SUITE.erl]: No you're right, leave it here.

File: src/gen_rpc_client.erl (hunk 6 above)
Comment (linearregression) [inline src/gen_rpc_client.erl]: Are you talking about get_goodnodes(Nodes) ?

**Comment by linearregression** [comment]
done

File: src/gen_rpc_client.erl (hunk 6 above)
Comment (linearregression) [inline src/gen_rpc_client.erl]: done

**Comment by linearregression** [comment]
please review.
